### ISSUE TITLE
[Feature Request] Possibility of skipped invoice numbers

### ISSUE THREAD

**Issue by danielcb** [frappe/erpnext#8765]
We're currently evaluating erpnext as a ERP (mainly for invoicing at the moment) and I noticed a small issue regarding invoice numbers / naming schemes:

We are using the following naming scheme for invoices: INV.YYYY.MM.#####. Assuming I've two booked invoices and two in a draft state:
Booked: INV20170500001
Booked: INV20170500002
Draft1: INV20170500003
Draft2: INV20170500004

If someone books/confirms Draft 2 the system creates INV20170500004. After that someone finds an issue in Draft1 and deletes it there will be an invoice number missing (INV20170500003).

At least in Germany this can be an issue. My understanding is that while it's not wrong/illegal (the requirement is mainly a unique numbering scheme) it certainly will be a red flag for an accountant/auditor. Other systems only create a number if an invoice gets confirmed/booked. This also guarantees invoice numbers in a "correct" order.

**Comment by nick9822** [comment]
@danielcb 
Instead of deleting you can cancel and amend invoice, so it will get revised.
Hope that helps.

You can post this on discuss.erpnext.com to get better insights.

**Comment by danielcb** [comment]
@nick9822 thanks for the quick follow up.

Yes, I've seen this, but its nonetheless possible to create "incomplete" invoice records this way. If the number would be assigned on confirmation this problem could be avoided completely. I've added it here since I thought it might end up in a backlog as enhancement somewhere. If the current solution is the preferred one, feel free to close this issue.

**Comment by tundebabzy** [comment]
Hi. I believe if there's any concern about maintaining the order of invoice numbers, the first thing to avoid is deleting invoices. Deleting invoices is also frowned at by auditors.

A work flow that avoids the need to delete invoices is to make use of sales orders. Once the details of the transaction have been confirmed, an invoice can be created from the finalized sales order. 

I'm summary, ERPNext is flexible enough that with relevant controls in place, the issue can be avoided.

**Comment by joezsweet** [comment]
Can we discuss this again?

**Comment by danielcb** [comment]
just my two cents:
Yes, deleting invoices is not a good idea, but imho, as long as they're not booked/confirmed, it doesn't matter. 

The proposed workaround indeed avoids the problem in most cases, but the problem is still there. I think invoice numbers (or any document IDs for that matter) should be added at the confirmation/booking stage and not after saving the document.

**Comment by alex-g-s** [comment]
My two cents:
- I use Sales orders when I am not sure when and if I will post the invoice. 
So put the sales data in Sales order, revise it as much as you need. You can post even deliveries and when everything is final - then adjust the sales order , submit it and then as follow up document Make an Invoice when ready and sure.

**Comment by Nayar** [comment]
> there's any concern about maintaining the order of invoice numbers, the first thing to avoid is deleting invoices. Deleting invoices is also frowned at by audit

No. The issue happens when 2 invoices are in draft mode and we submit another invoice. Even if we cancel the draft invoice, we got ourselves a hole
